src/MongoDB.Driver.Core/Core/Configuration/ClusterSettings.cs

```diff
@@ -251,6 +284,87 @@ public IServerSelector PostServerSelector
                 postServerSelector: Optional.Create(postServerSelector.WithDefault(_postServerSelector)),
                 schemaMap: Optional.Create(schemaMap.WithDefault(_schemaMap)),
                 scheme: scheme.WithDefault(_scheme));
+            clusterSettings.UpdateConnectionModeWasSet(connectionMode.HasValue || _connectionModeWasSet);
+            return clusterSettings;
+        }
+
+        /// <summary>
+        /// Returns a new ClusterSettings instance with connection settings changed.
+        /// </summary>
+        /// <param name="connectionMode">The connection mode.</param>
+        /// <param name="directConnection">The direct connection.</param>
+        /// <returns></returns>
+        [Obsolete("Use With instead.")]
+        public ClusterSettings WithConnection(ClusterConnectionMode connectionMode, bool? directConnection)
+        {
+            var clusterSettings = new ClusterSettings(
+                endPoints: Optional.Enumerable(_endPoints),
+                kmsProviders: Optional.Create(_kmsProviders),
+                localThreshold: _localThreshold,
+                maxServerSelectionWaitQueueSize: _maxServerSelectionWaitQueueSize,
+                replicaSetName: _replicaSetName,
+                serverSelectionTimeout: _serverSelectionTimeout,
+                preServerSelector: Optional.Create(_preServerSelector),
+                postServerSelector: Optional.Create(_postServerSelector),
+                schemaMap: Optional.Create(_schemaMap),
+                scheme: _scheme);
+            clusterSettings.UpdateConnectionModeWasSet(!directConnection.HasValue || _connectionModeWasSet);
+
+            if (directConnection.HasValue)
```

Comment:
this is the main idea of this method: We will set `connectionMode` only in case if DirectConnection is not set.